File: packages/protopy/protofy/agent/ProtoAgentsInterface.py

from abc import ABC, abstractmethod 
from typing import TypeVar, Type
import json
import os


# T being any class that implements "ProtoAgentInterface"
T = TypeVar('T', bound='ProtoAgentInterface')

class ProtoAgentInterface(ABC): 
    @abstractmethod
    def configure(self: T, subsystems) -> T: 
        pass

    def configure_from_file(self, path): 
        abs_path = os.path.join(os.getcwd(), path)
        
        with open(abs_path, 'r') as file:
            subsystems = json.load(file)
        return self.configure(subsystems)

    @abstractmethod
    def connect(self, mqtt_host, mqtt_port): 
        pass

    @abstractmethod
    def pub_monitor(self, subsystem_name,  monitor_name, value): 
        pass

    @abstractmethod
    def handle(self, subsystem_name, action_name, handler): 
        pass

File: packages/agents/js/ProtoAgentsInterface.ts
import fs from 'fs'
import path from 'path'

export class ProtoAgentInterface {
    constructor() {
        if (this.constructor === ProtoAgentInterface) {
            throw new Error('Cannot instantiate abstract class ProtoAgentInterface');
        }
    }

    configure(subsystems) {
        throw new Error('Method "configure" must be implemented by subclasses');
    }

    configureFromFile(jsonPath) {
        const absPath = path.join(process.cwd(), jsonPath);

        const fileContents = fs.readFileSync(absPath, 'utf8');
        const subsystems = JSON.parse(fileContents);
        return this.configure(subsystems);
    }

    connect(...any) {
        throw new Error('Method "connect" must be implemented by subclasses');
    }

    pubMonitor(subsystemName, monitorName, value) {
        throw new Error('Method "pubMonitor" must be implemented by subclasses');
    }

    handle(subsystemName, actionName, handler) {
        throw new Error('Method "handle" must be implemented by subclasses');
    }
}